# Errors deploying more than one key vault

```
arm-deployment: set-resource-group-name set-storage-account-name set-azure-template-tag set-azure-account set-azure-resource-group-tags set-key-vault-names ## deploy container/kv to store terraform state for each environment
	az deployment sub create --name "resourcedeploy-tsc-$(shell date +%Y%m%d%H%M%S)" \
		-l "UK South" --template-uri "https://raw.githubusercontent.com/DFE-Digital/tra-shared-services/9b29f223e5b0019a2bc94cb1653e2b74a9934742/azure/resourcedeploy.json" \
		--parameters "resourceGroupName=${RESOURCE_GROUP_NAME}" 'tags=${RG_TAGS}' \
			"tfStorageAccountName=${STORAGE_ACCOUNT_NAME}" "tfStorageContainerName=${SERVICE_SHORT}-tfstate" \
			keyVaultName="${KEY_VAULT_APPLICATION_NAME}", "${KEY_VAULT_INFRASTRUCTURE_NAME}")' \
			"location=uksouth" "logAnalyticsWorkspaceName=s189d01-rtt-rv-kv-log" "enableDiagnostics=true" \
			"enableKVPurgeProtection=${KEY_VAULT_PURGE_PROTECTION}" ${WHAT_IF}
```
Will generate an error because more than one key vault is specified. If set to single KV it works fine.

Error code is:
```
MultipleErrorsOccurred - Multiple error occurred: BadRequest,BadRequest. Please see details.
BadRequest - Deployment template validation failed: 'The value for the template parameter 'logAnalyticsWorkspaceName' at line '20' and column '34' is not provided. Please see https://aka.ms/arm-create-parameter-file for usage details.'.
BadRequest - Deployment template validation failed: 'The value for the template parameter 'logAnalyticsWorkspaceName' at line '20' and column '34' is not provided. Please see https://aka.ms/arm-create-parameter-file for usage details.'.
make: *** [Makefile:257: arm-deployment] Error 1
```

@saliceti @mrzta

--- a/azure/resourcedeploy.json
+++ b/azure/resourcedeploy.json
@@ -2,13 +2,14 @@
   "$schema": "https://schema.management.azure.com/schemas/2018-05-01/subscriptionDeploymentTemplate.json#",
   "contentVersion": "1.0.0.0",
   "parameters": {
+    "keyVaultNames": {
+      "type": "array",
+      "defaultValue": [],
+      "optional": true
+    },
     "keyVaultName": {
       "type": "string",
       "defaultValue": ""
-    },
-    "keyVaultNames": {
-      "type": "array",
-      "defaultValue": []
     },
     "enableKVPurgeProtection": {
       "type": "bool",
@@ -34,16 +35,25 @@
     "tags": {
       "type": "object"
     },
-    "location": {
-      "type": "string",
-      "defaultValue": "[deployment().location]",
-      "metadata": {
-        "description": "Location for the deployments and the resources"
-      }
-    },
     "utcValue": {
       "type": "string",
       "defaultValue": "[utcNow('yyMMddHHmmss')]"
+    },
+    "sku": {
+      "type": "string",
+      "defaultValue": "PerGB2018"
+    },
+    "location": {
+      "type": "string",
+      "defaultValue": "uksouth"
+    },
+    "enableDiagnostics": {
+      "type": "bool",
+      "defaultValue": false
+    },
+    "logAnalyticsWorkspaceName": {
+      "type": "string",
+      "optional": true
     }
   },
   "variables": {
@@ -82,7 +92,9 @@
           }
         }
       },
-      "dependsOn": ["[parameters('resourceGroupName')]"]
+      "dependsOn": [
+        "[parameters('resourceGroupName')]"
+      ]
     },
     {
       "type": "Microsoft.Resources/deployments",
@@ -131,10 +143,18 @@
           },
           "enablePurgeProtection": {
             "value": "[parameters('enableKVPurgeProtection')]"
+          },
+          "enableDiagnostics": {
+            "value": "[parameters('enableDiagnostics')]"
+          },
+          "logAnalyticsWorkspaceName": {
+            "value": "[parameters('logAnalyticsWorkspaceName')]"
           }
         }
       },
-      "dependsOn": ["[parameters('resourceGroupName')]"]
+      "dependsOn": [
+        "[parameters('resourceGroupName')]"
+      ]
     },
     {
       "type": "Microsoft.Resources/deployments",
@@ -162,10 +182,18 @@
           },
           "enablePurgeProtection": {
             "value": "[parameters('enableKVPurgeProtection')]"
+          },
+          "enableDiagnostics": {
+            "value": "[parameters('enableDiagnostics')]"
+          },
+          "logAnalyticsWorkspaceName": {
+            "value": "[parameters('logAnalyticsWorkspaceName')]"
           }
         }
       },
-      "dependsOn": ["[parameters('resourceGroupName')]"]
+      "dependsOn": [
+        "[parameters('resourceGroupName')]"
+      ]
     },
     {
       "type": "Microsoft.Resources/deployments",
@@ -191,7 +219,33 @@
           }
         }
       },
-      "dependsOn": ["[parameters('resourceGroupName')]"]
+      "dependsOn": [
+        "[parameters('resourceGroupName')]"
+      ]
+    },
+    {
+      "condition": "[parameters('enableDiagnostics')]",
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2020-10-01",
+      "name": "[format('create-{0}-account-{1}', parameters('logAnalyticsWorkspaceName'), parameters('utcValue'))]",
+      "resourceGroup": "[parameters('resourceGroupName')]",
+      "properties": {
+        "mode": "Incremental",
+        "templateLink": {
+          "relativePath": "templates/log-analytics.json"
+        },
+        "parameters": {
+          "logAnalyticsWorkspaceName": {
+            "value": "[parameters('logAnalyticsWorkspaceName')]"
+          },
+          "sku": {
+            "value": "[parameters('sku')]"
+          },
+          "location": {
+            "value": "[parameters('location')]"
+          }
+        }
+      }
     },
     {
       "type": "Microsoft.Resources/deployments",

--- a/azure/templates/key-vault.json
+++ b/azure/templates/key-vault.json
@@ -3,17 +3,27 @@
   "contentVersion": "1.0.0.0",
   "parameters": {
     "location": {
-      "type": "string"
+      "type": "string",
+      "defaultValue": "uksouth"
     },
     "keyVaultName": {
       "type": "string"
     },
     "enabledForTemplateDeployment": {
-      "type": "bool"
+      "type": "bool",
+      "defaultValue": true
     },
     "enablePurgeProtection": {
       "type": "bool",
       "defaultValue": true
+    },
+    "logAnalyticsWorkspaceName": {
+      "type": "string",
+      "optional": true
+    },
+    "enableDiagnostics": {
+      "type": "bool",
+      "defaultValue": false
     }
   },
   "variables": {
@@ -35,6 +45,26 @@
           "family": "A",
           "name": "standard"
         }
+      }
+    },
+    {
+      "condition": "[parameters('enableDiagnostics')]",
+      "type": "Microsoft.KeyVault/vaults/providers/diagnosticsettings",
+      "apiVersion": "2017-05-01-preview",
+      "name": "[format('{0}/Microsoft.Insights/diagnosticSettings', parameters('keyVaultName'))]",
+      "properties": {
+        "name": "keyVaultDiagnosticSettings",
+        "workspaceId": "[resourceId('Microsoft.OperationalInsights/workspaces', parameters('logAnalyticsWorkspaceName'))]",
+        "logs": [
+          {
+            "category": "AuditEvent",
+            "enabled": true
+          },
+          {
+            "category": "AzurePolicyEvaluationDetails",
+            "enabled": true
+          }
+        ]
       }
     }
   ]

--- a/azure/templates/log-analytics.json
+++ b/azure/templates/log-analytics.json
@@ -1,0 +1,30 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "logAnalyticsWorkspaceName": {
+      "type": "string"
+    },
+    "sku": {
+      "type": "string",
+      "defaultValue": "PerGB2018"
+    },
+    "location": {
+      "type": "string",
+      "defaultValue": "uksouth"
+    }
+  },
+  "resources": [
+    {
+      "type": "Microsoft.OperationalInsights/workspaces",
+      "apiVersion": "2021-12-01-preview",
+      "name": "[parameters('logAnalyticsWorkspaceName')]",
+      "location": "[parameters('location')]",
+      "properties": {
+        "sku": {
+          "name": "[parameters('sku')]"
+        }
+      }
+    }
+  ]
+}

